### DISCUSSION TITLE
Deploying CoreDNS to Amazon EKS using IAM Role for Service Account (IRSA)

### DISCUSSION THREAD

**Discussion by vijayansarathy** [coredns/coredns#5149] [General]
I am trying to deploy the stock CoreDNS deployment in an EKS cluster to run under the identity of a service account mapped to an IAM role using [IAM Roles for Service Accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) feature. I have mapped the **coredns** service account to an IAM role with the following annotations on the service account:

```
  annotations:
    eks.amazonaws.com/component: coredns
    eks.amazonaws.com/role-arn: arn:aws:iam::XXXXXXXX:role/EKS-CoreDNS-ServiceAccount-Role
    k8s-app: kube-dns
```

The IAM role is associated with the permission policy below that grants access to Route53.

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "route53:ListHostedZonesByName",
                "route53:ListResourceRecordSets"
            ],
            "Resource": "*"
        }
    ]
}
```
The **CoreFile** field in **coredns** ConfigMap look as follows:

```
  Corefile: |
    .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
    mydomain.com {
      route53 mydomain.com.:MYDOMAIN-HOSTED-ZONEID {
      }
    }
```

However, when the Route53 plugin gets loaded, it is going to the IAM role associated with the EC2 instance profile rather than the one associated with **coredns** role. Has anyone successfully deployed CoreDNS using IRSA?

Here's the error I see in the logs of the CoreDNS pods:

```
plugin/route53: /etc/coredns/Corefile:17 - Error during parsing: failed to create route53 plugin: AccessDenied: User: arn:aws:sts::XXXXX:assumed-role/eks-prod-managed-workers-st-WorkerNodeInstanceRole-1QDJDINQZBFMI/i-0947b092115106035 is not authorized to perform: route53:ListHostedZonesByName because no identity-based policy allows the route53:ListHostedZonesByName action
	status code: 403, request id: 9325c5ac-79c4-484d-892c-191eff8466dc
```

**Comment by chrisohaver** [comment]
There's now an open issue/bug for this #5153

**Reply by spstarr** [reply]
I just hit this issue too... looking forward to fix

**Comment by chrisohaver** [comment]
There is a proposed fix opened (https://github.com/coredns/coredns/pull/5268).
